MRG: Add dpss window function

larsoner opened the pull request:
Takes care of the DPSS half of #2221.
Closes #4354.

Needs:

* [x] Use `eig_banded`
* [x] more tests
* [x] check that all issues in #4354 are covered
* [x] Deprecate `slepian`
* [x] Add `See Also` links
* [x] Add `.. versionadded::`
* [x] Use LAPACK `stebz`/`stein` for tridiagonal ~~Cythonize the tridiagonal code~~
* [x] Decide the best way to do the lobe check
* [x] change docstring once #7898 / #7900 is in

larsoner commented:
@arokem do you remember why you did `linalg.eigvals_banded` followed by a custom tridiagonal solver to get the eigenvectors instead of using `linalg.eig_banded` directly? This seems to have been done [intentionally back in 2011](https://github.com/nipy/nitime/commit/6afdc1ba466cd2c7c5a839e2b0244ce6067f867f#diff-fedde05cdadb91eccad51064f81818f8L422) I assume for performance reasons (?), is this less of a concern now that windows can be interpolated?

arokem commented:
I *believe* so, but maybe @miketrumpis can comment? He wrote that code. 

I think that performance might still be a concern, even with interpolation in place, but honestly, it's been a while...

There is a cythonized version of the tridiagonal solver (also Mike's work!) here: https://github.com/nipy/nitime/blob/master/nitime/_utils.pyx. IIRC, using that was a big win performance-wise.

miketrumpis commented:
TL;DR -- with Cythonized inverse iteration, I am in practice calculating DPSS with lengths of 2^10-2^16 with no painfully obvious slowdown.  I can vouch for it being pretty fast ;)

I don't perfectly recall all the moving pieces regarding inverse iteration versus interpolation, as we implemented these at about the same time. In my own use, I implemented inverse iteration and never thought twice about interpolation. Further, I never was assured (I mean I never looked) whether interpolated DPSS remain orthonormal, which is important for certain DPSS applications and relevant to the "approximately uncorrelated" claim of multitaper spectral estimations.

The "[inverse iteration](https://en.wikipedia.org/wiki/Inverse_iteration)" solver is a quick way of finding eigenvectors when you already have an eigenvalue, provided by eigvals_banded. Doing a quick test just now for a banded 10000x10000 matrix: computation for 20 eigenvalues takes 64 ms, and subsequent estimation of eigenvectors is (well less than) 10 ms per vector. On the other hand, using eig_banded() to compute 20 eigen-values/vectors takes 1.4 s (old macbook pro w/ vecLib). 

I believe, as written, we are solving an alternative matrix factorization that gives identical eigenvectors to the true DPSS problem, but with different eigenvalues. Once the "true" eigenvectors are estimated, the true eigenvalues can be computed, but I'd have to go back to Perceval and Walden to remember the logic behind that particular computation.

pv commented:
Lapack iirc does have specialized tridiagonal eigensolvers.

29.8.2017 6.53 "miketrumpis" <notifications@github.com> kirjoitti:

> TL;DR -- with Cythonized inverse iteration, I am in practice calculating
> DPSS with lengths of 2^10-2^16 with no painfully obvious slowdown. I can
> vouch for it being pretty fast ;)
>
> I don't perfectly recall all the moving pieces regarding inverse iteration
> versus interpolation, as we implemented these at about the same time. In my
> own use, I implemented inverse iteration and never thought twice about
> interpolation. Further, I never was assured (I mean I never looked) whether
> interpolated DPSS remain orthonormal, which is important for certain DPSS
> applications and relevant to the "approximately uncorrelated" claim of
> multitaper spectral estimations.
>
> The "inverse iteration <https://en.wikipedia.org/wiki/Inverse_iteration>"
> solver is a quick way of finding eigenvectors when you already have an
> eigenvalue, provided by eigvals_banded. Doing a quick test just now for a
> banded 10000x10000 matrix: computation for 20 eigenvalues takes 64 ms, and
> subsequent estimation of eigenvectors is (well less than) 10 ms per vector.
> On the other hand, using eig_banded() to compute 20 eigen-values/vectors
> takes 1.4 s (old macbook pro w/ vecLib).
>
> I believe, as written, we are solving an alternative matrix factorization
> that gives identical eigenvectors to the true DPSS problem, but with
> different eigenvalues. Once the "true" eigenvectors are estimated, the true
> eigenvalues can be computed, but I'd have to go back to Perceval and Walden
> to remember the logic behind that particular computation.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/7802#issuecomment-325548053>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5pP1Djjkeav2_-PHupxpPK3Zhnx1ks5sc4s6gaJpZM4PFANx>
> .
>

miketrumpis commented:
Good point -- it looks like in principle the same two step eigenvalue -> inverse iteration can be done with LAPACK.

http://www.netlib.org/lapack/explore-3.1.1-html/dstebz.f.html
http://www.netlib.org/lapack/explore-3.1.1-html/dstein.f.html

But I'm not yet finding those routines wrapped by scipy.linalg.lapack (version 0.19.0).  For DPSS purposes it would be overkill to solve the whole spectral decomposition, since you almost never want more than a small fraction of eigenvectors.

larsoner commented:
> But I'm not yet finding those routines wrapped by scipy.linalg.lapack 

That's fine, I can add them separately or as part of this.

> For DPSS purposes it would be overkill to solve the whole spectral decomposition, since you almost never want more than a small fraction of eigenvectors.

If we use `dstebz` to find the limited eigenvalues of interest, we can then use `dstein` with those specific values to avoid any unnecessary computations, though, right?

miketrumpis commented:
Yes, I think you could bypass the nitime `tridi_inverse_iteration` and `tridisolve` if you wanted, avoiding a heavy Cython patch. If you wrap the LAPACK, I'd be happy to help adapt the dpss routine. Should be easy if the linalg is working.

larsoner commented:
> If you wrap the LAPACK, I'd be happy to help adapt the dpss routine.

Based on the code I think it's straightforward enough that I can do it. It should actually simplify / remove a few lines because you provide the diagonal and off-diagonal vectors directly instead of packing them. 

I'll work on the LAPACK stuff first, then once that's in come back to this. Thanks for helping sort out the fastest way to do it.

I wonder if we even need the interpolation code once the LAPACK routines are in place. I seem to remember the interpolation being necessary when using *really* long windows (e.g., 1e6 elements), but maybe such applications should be rare enough that we won't need it, at least for a first pass. I'm leaning toward removing it for now -- callers can always interpolate if they see a need to. @arokem @miketrumpis okay with you?

arokem commented:
Fine by me. I believe that the interpolation trick historically preceded the Cython implementation.

miketrumpis commented:
Good call to leave interpolation to the user.

miketrumpis commented:
I am just playing around with the new `eig_tridiagonal` function in `scipy.linalg.decomp`.  Looks like the same eigenvalues/vectors are returned.  If I manually set the correct `lapack_driver` option, then speed is comparable too (a blink faster than the Cython).

larsoner commented:
I meant to do some benchmarks, did you try both `stebz` and `stemr`?

miketrumpis commented:
Accidentally, yes.
Actually, I'd recommend leaving `lapack_driver='stbez'` explicit in any DPSS calculation in case of intentional or accidental future API snafu. Using `stemr` was nearly 10x slower and crashed my memory (I assume it solves the whole system?).

larsoner commented:
Great, will do. What code crashes it on your system? I'd like to try locally, too.

miketrumpis commented:
Apologies for the misleading word "crash", I just had to kill the process b/c it suddenly cached out about 10 gb to disk. I think the source of the problem is a bug in auto-selecting the lapack driver, noted on the scipy pull request.

larsoner commented:
@miketrumpis we have a problem -- the `max(abs(...))` algorithm from nitime that implements the "first lobe positive" criteria for the odd tapers is susceptible to numerical noise. It passes on my machine, but fails for Travis on different tapers for different runs (see logs). Looking at the tapers for these failures, `max(w)` is very close to `min(w)`, so `max(abs(w))` isn't robust.

WDYT about just doing this:
```
    # * antisymmetric tapers should begin with a positive lobe
    #   (this depends on the definition of "lobe", here we'll take the first
    #   point above the numerical noise, which should be good enough for
    #   sufficiently smooth functions, and more robust than relying on an
    #   algorithm that uses max(abs(w)), which is susceptible to numerical
    #   noise problems)
    for i, w in enumerate(windows[1::2]):
        if w[np.abs(w) > 1e-7][0] < 0:
            windows[2 * i + 1] *= -1
```

larsoner commented:
Okay I think this is ready for review. I still need to swap in the tridiagonal solver (#7810) once it's in, but that's simple. @endolith @rgommers in particular you might be interested because you were involved a lot in #4354.

Example here:

https://4105-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.dpss.html#scipy.signal.dpss

It would be great to have this for 1.0, but the timeline is probably too tight for that.

endolith commented on scipy/signal/windows.py:
```diff
@@ -1473,6 +1480,8 @@ def slepian(M, width, sym=True):
     >>> plt.xlabel("Normalized frequency [cycles per sample]")
 
     """
+    warnings.warn('slepian is deprecated and will be removed in a future '
```
docstring should say it's deprecated, too

larsoner commented on scipy/signal/dpss.py:
```diff
@@ -0,0 +1,181 @@
+    Compute the Discrete Prolate Spheroidal Sequences (DPSS).
+
+    This uses the tridiagonal formulation given in [1]_.
+
```
(multiple) DPSS windows are used a lot in multitaper analyses

larsoner commented on scipy/signal/dpss.py:
```diff
@@ -0,0 +1,181 @@
+    >>> axes[2, 1].legend(['DPSS', 'Kaiser'])
+    >>> fig.tight_layout()
+
+    """  # noqa: E501
```
Yes it works for PEP8 checkers / linters (for multi-line strings it should come at the end, which makes sense because anywhere else and it would be part of the string)

endolith replied:
oh, it's for the URL in the docstring?

larsoner replied:
Yep

larsoner commented:
Comments addressed, tweaked docs here (another example, better var names and descriptions):

https://4111-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.dpss.html#scipy.signal.dpss

miketrumpis commented:
L2 normalization and inverse iteration work as I'd expect if I naively call `dpss`, so I'm happy.  Chalk this up to ignorance, but I'm confused by the use of symmetric/period windows in this case.  Any rate, the default `sym=True` gives what I have thought were correct windows.

larsoner commented:
> L2 normalization and inverse iteration work as I'd expect if I naively call dpss, so I'm happy

Actually the default is to normalize to the max if you just call `dpss(a, b)` as with other window functions. L2 norm happens when you request some number of windows.

larsoner commented:
@endolith if you have time to look, are you happy with this one now?

endolith commented on scipy/signal/windows.py:
```diff
@@ -1639,6 +1652,217 @@ def exponential(M, center=None, tau=1., sym=True):
+
+    Examples
+    --------
+    We can compare the window to `dpss`, which was invented as an alternative
```
"compare the window to `kaiser`"

endolith commented:
Hmm, you said this fixes https://github.com/scipy/scipy/issues/2221 but it doesn't seem to?

```
dpss(4, 1)
Out[31]: array([ 0.535,  1.   ,  1.   ,  0.535])

hann(4)
Out[33]: array([ 0.  ,  0.75,  0.75,  0.  ])
```

It looks like it does the same thing that slepian did:

```
    if norm == 'inf':
        windows /= windows.max()
```

larsoner commented:
> I'm not sure what the "correct" answer is, but the energy of the windows would change in a staggered way for even vs odd length when you just normalize to max value.

Argh, you're right.

Unless someone knows the answer, we might be stuck with this normalization as as close as we can get for now. So how about:

1. Remove the "closes" label
2. Add a note about the double-1 value for even-length windows
3. If we figure out a better way to do it in the future, we can add a new normalization mode.

WDYT? The drawback I can see with this approach is that it means that our default normalization for e.g. `get_window` might be less than ideal in the future. But it could be fixed via `kwarg` if we ever figure out how to do it.

endolith commented:
I'm ok with that.  `dpss` still improves on `slepian` in other ways, right?  It's not a regression to leave a bug unfixed.

> our default normalization for e.g. get_window might be less than ideal in the future

I don't know what you mean by that

> But it could be fixed via kwarg if we ever figure out how to do it.

I don't know what you mean by that, either.  Don't we just leave it as is for now, and then later the output signal amplitude changes a little when we fix the normalization?

larsoner commented:
> Don't we just leave it as is for now, and then later the output signal amplitude changes a little when we fix the normalization?

Technically this would be a backward compatibility break.

>> it could be fixed via kwarg if we ever figure out how to do it.

I mean that we can add a new mode, like `norm='window'` that is the same as `inf` for odd windows and like other windows (no 1 values at all) for even windows. Then people using `get_window` just need to provide the extra argument to get the more standard behavior.

rgommers commented:
This one is not going to make the cut unfortunately, bumping to 1.1.0

larsoner commented:
@miketrumpis any ideas about how to improve the lobe check / how to properly fit the textbook definition? (Or do we not really need to?)

@endolith are you okay with using `norm='inf'` the default for now (which is correct mathematically but inconsistent with other window functions) and leaving open the future possibility of `norm='window'` or so for if/when we figure out how to make even-length windows not have any 1 values? If not I'm not sure what the path forward would be.

Rebased and changed to use `eigh_tridiagonal`, so this should be good to go once we sort out these last few issues.

endolith commented:
> are you okay with using norm='inf' the default for now

Yes

> and leaving open the future possibility of norm='window' or so for if/when we figure out how to make even-length windows not have any 1 values?

Well, I'd be fine with the `norm='inf'` just being documented as slightly wrong for even lengths, with a note that it will hopefully be fixed at a later date.  If the note is there from the beginning of the function, is it that bad to change it later?  It's only a fraction of a decibel error, and `slepian` and `chebwin` behaved this way for a long time.

![dpss energy vs n](https://user-images.githubusercontent.com/58611/30529132-3fd3ab36-9c08-11e7-97db-f028809aa677.png)

What do the words `inf` and `l2` mean, by the way?

larsoner commented:
> Well, I'd be fine with the norm='inf' just being documented as slightly wrong for even lengths, with a note that it will hopefully be fixed at a later date.

I'm not sure if we can do it by default, even if it's a small change. That's why I propose using a different `norm='window'` that did the correction if we ever figure it out.

> What do the words inf and l2 mean, by the way?

[l2 norm or l-inf norm](https://docs.scipy.org/doc/scipy-0.19.1/reference/generated/scipy.linalg.norm.html). Speaking of which, I'll change it to `2` instead of `l2` for consistency.

miketrumpis commented:
I think I understood the potential lobe-check bug in principle, but for what DPSS parameterization does the proposed method fail?  I had actually not checked more rigorously than by eye for several parameter sets.  As you allude, the sign of the tapers does not matter computationally in any way I'm aware of.

larsoner commented:
It failed for a number of tests I ran locally, but I originally found the problem in the Travis logs. I tried tweaking the slope code, but it wasn't working well. Does using the sign of the first point above some reasonable estimate of numerical noise (e.g. 1e-7) seem reasonable to you?

larsoner commented:
@endolith I think I empirically derived a correction factor that is reasonable for the even-M cases. Can you look at the updated docstring and example [here](https://4317-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.dpss.html#scipy.signal.dpss) and see if you are convinced?

miketrumpis commented:
Perhaps safer: sign of the first value above the uniform level of squared values (1/M).  Or rather max(1e-7, 1/M). Forgive some very compressed syntax:

```python
tol = max(1e-7, M**-1.0)
s = np.array([np.sign(dpss[i, np.where(dpss[i]**2 > tol)[0][0]]) for i in xrange(1, len(dpss), 2)])
dpss[1::2] *= s[:, np.newaxis]
```

larsoner commented:
That seems reasonable to me, will make that change.

When you get a chance can you give the PR another round of review? Other
than adding the max in your code snippet I think this should be good to go.

rgommers commented:
@larsoner did you see [this thread](https://mail.python.org/pipermail/scipy-dev/2017-January/021654.html)? We decided there to not add more windows to the ``signal`` namespace but make ``signal.windows`` public. I didn't actually touch those PRs after that - maybe this one is a good time to start.

larsoner commented:
Sure, I can move it.

miketrumpis commented:
So regarding issue #2221, isn't it simple to calculate the DPSS with length `(M//2) * 2 + 1` and then pop out the middle value if even-M was requested?

larsoner commented:
[No](https://gist.github.com/larsoner/6b4e8d718db7f5651437a3efc8ad494e) (feel free to double-check I did what you meant)
![figure_1](https://user-images.githubusercontent.com/2365790/30615071-a2f875e4-9d5b-11e7-8975-8b2d08139273.png)

larsoner commented:
(I mean yes it's simple, no it doesn't work)

endolith commented:
No, that produces a different sequence.

You can find the fft interpolated peak though, by taking the mean of the
spectrum after multiplying by complex exponential, to effectively shift the
peak to zero position.
Not sure if this is the true peak or not



On Sep 19, 2017 4:40 PM, "miketrumpis" <notifications@github.com> wrote:

So regarding issue #2221 <https://github.com/scipy/scipy/issues/2221>,
isn't it simple to calculate the DPSS with length (M//2) * 2 + 1 and then
pop out the middle value if even-M was requested?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/scipy/scipy/pull/7802#issuecomment-330666765>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AADk8_Hb3hgIIeqoAg54R_4KwLg8Im4wks5skCalgaJpZM4PFANx>
.

larsoner commented:
Do you have time to check if interpolation does better than the existing
empirically derived correction factor?

endolith commented:
I'm not sure how this normalization should work with the multi-tapers, but for the zeroth-order taper, instead of:

    windows *= M**2 / float(M**2 + NW)

it would be:

    s = fftpack.fft(window)
    k = np.concatenate((np.arange(M/2), np.arange(-M/2, 0)))
    shift = -(M/2 - 0.5)
    s_sh = s * np.exp(-1j * 2 * np.pi * shift / M * k)
    window /= s_sh.mean().real

The fft and exp both add computation time, though.  Maybe it can be simplified, though?

larsoner commented:
Nice -- it looks like [it works](https://gist.github.com/larsoner/488d3c2049986dc87b1800727c6b920c) and is in general quite close to the approximate correction:

![figure_1](https://user-images.githubusercontent.com/2365790/30625882-1f974e08-9d94-11e7-8cc0-d74d70a470f9.png)

Maybe we should have `norm='approximate'` and `norm='fft'` for these two modes? The former by default because it's fast and probably good enough almost all the time, and the latter for enhanced precision if people want/need it?

endolith commented:
Hmm..  There are big discrepancies at low N?

I'm not even sure what the "correct" behavior is.  

When used in filter design, it looks like things like `scipy.signal.firwin` ignore this normalization anyway, and re-normalize so that sum of samples = 1 for unity gain at DC.  

For multi-taper spectral estimation with meaningful numbers (compensating for "coherent power gain" or "processing gain"?), you would use the `l2` norm, which puts the spectrum peak at 0 dB?

larsoner commented:
> There are big discrepancies at low N?

Only for 2 and 4 samples. The existing correction seems to do a little bit better there, the FFT one a little bit better at higher N (less up-down fluctuation).

And I agree for filter design and multitaper estimation it shouldn't matter. So maybe the approximate one is good enough for now.

endolith commented:
Well, 2 sample windows are basically nonsense, and the other window functions handle them inconsistently.

I don't really have a strong opinion on which to use.  I tried to plug in a bunch of peak values to zunzun.com to see if it could find a formula for `peak_L2_value(M, NW)`, but it couldn't.

larsoner commented:
@endolith CircleCI just finished building the latest iteration which includes both the subsample/FFT correction and (default) empirical correction, can you see if you're happy with the wording/examples:

https://4387-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.signal.windows.dpss.html#scipy.signal.windows.dpss

endolith commented:
You still need to fix the example text, example is comparing dpss to kaiser, not to itself.

> For maximum speed

This seems kind of irrelevant, since the L2 norm is so different.  Maybe "for the orthonormal case used in multitaper analysis, use '2'"?  Also you can still use 'inf' for maximum speed if you don't care about the normalization, right?

Otherwise it looks fine to me.

larsoner commented:
> You still need to fix the example text, example is comparing dpss to kaiser, not to itself.

The first example, yes -- but the third compares `dpss` to itself with three normalizations (l-inf, subsample, approximate) modes. Did you see that? If you did, then I'm not sure what you mean here.

larsoner commented:
> Also you can still use 'inf' for maximum speed if you don't care about the normalization, right?

No it's still slower than the l2 norm. The original computation gives orthonormal vectors, so any other normalization (the `max()` done by l-infinity norm included) will be slower because it must be done on top of this. I can clarify this in the docs, though.

endolith commented:
> The first example, yes

Yes, I mean this part:

"We can compare the window to dpss, which was invented as an alternative that was easier to calculate"

should become

"We can compare the window to kaiser, which was invented as an alternative that was easier to calculate"

> No it's still slower than the l2 norm.

Ok, but it's a very minor difference in speed, no?  Only the fft method is significantly slower.  I think the fact that the normalization is not even close to 1 is much more important here.

larsoner commented:
Good points, fixed in latest commit (force-pushed).

It sounds like @miketrumpis and @endolith are happy with this version (please correct me if I'm wrong). Anyone else who is interested, feel free to have a look.

pv commented on scipy/signal/windows/windows.py:
```diff
@@ -1423,6 +1425,11 @@ def slepian(M, width, sym=True):
     Used to maximize the energy concentration in the main lobe.  Also called
     the digital prolate spheroidal sequence (DPSS).
 
+    .. note:: Deprecated in SciPy 1.0.
```
1.1

larsoner commented:
Thanks @pv, fixed.

pv commented:
Ok, no further comments apparently and discussion appears converged, so merging.
